File: solutions/python3/0001.two-sum/solution.py
# Created by seognil LC at 2023/07/09 22:12
# leetgo: 1.3.3
# https://leetcode.cn/problems/two-sum/

from typing import *
from leetgo_py import *

# * [ '40 ms', '87%', '17.3 MB', '27%' ]

# @lc code=begin


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        idx_map = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in idx_map:
                return [idx_map[complement], i]
            idx_map[num] = i


# @lc code=end

if __name__ == "__main__":
    nums: List[int] = deserialize("List[int]", read_line())
    target: int = deserialize("int", read_line())
    ans = Solution().twoSum(nums, target)

    print("\noutput:", serialize(ans))

File: solutions/rust/src/0001.two-sum/solution.rs
// Created by seognil LC at 2023/07/09 21:48
// leetgo: 1.3.3
// https://leetcode.cn/problems/two-sum/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// * [ '0 ms', '100%', '2.6 MB', '6%' ]

// @lc code=begin

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        use std::collections::HashMap;
        let mut map = HashMap::new();
        for (i, &num) in nums.iter().enumerate() {
            if let Some(&j) = map.get(&(target - num)) {
                return vec![j as i32, i as i32];
            }
            map.insert(num, i);
        }
        vec![]
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let target: i32 = deserialize(&read_line()?)?;
    let ans: Vec<i32> = Solution::two_sum(nums, target).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
